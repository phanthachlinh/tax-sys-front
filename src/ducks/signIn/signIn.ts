import axios from 'axios';

export default function(state:IUser|null=null,action:any){
  switch (action.type){
    case(SignInTypes.GET_USER):
      if(!action.payload.data)
        return 'fail'
      else
        return {ID:action.payload.data.ID,isManager: action.payload.data.isManager}
    case(SignInTypes.LOG_OUT):
      return null
    default:
        return state
  }
}
export const userActions = {
  signIn: (username:String,password:String)=>{
    return {
      type: SignInTypes.GET_USER,
      payload: axios.get('http://127.0.0.1:8001/user/validate?username='+username+'&password='+password+''),
      meta:{
        username: username
      }
    }
  },
  logout: ()=>{
    return {
      type: SignInTypes.LOG_OUT,
      payload: {}
    }
  }
}
export interface IUser{
  ID: number,
  isManager: boolean
}

export const SignInTypes = {
  GET_USER: 'GET_USER',
  LOG_OUT: 'LOG_OUT'
}
