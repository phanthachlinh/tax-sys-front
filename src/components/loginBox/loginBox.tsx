import React,{Component} from 'react';
import styled from 'styled-components';
import {connect} from 'react-redux';
import { userActions, IUser } from '../../ducks/signIn/signIn';
import { Redirect } from 'react-router-dom';
export class LoginBox extends Component<IProps,IState>{
    constructor(props:IProps){
      super(props);
      this.state={
        username: '',
        password:'',
        loginError:false,
        isUserLoggedIn:false
      }
    }
   componentDidMount(){
     if(this.props.user!==null&&!this.state.isUserLoggedIn)
      this.setState({isUserLoggedIn:true})

   }
   componentDidUpdate(){
     if(this.props.user!==null&&!this.state.isUserLoggedIn)
      this.setState({isUserLoggedIn:true})
   }
    changeUsernameHandler(ev:any){
      this.setState({username:ev.target.value});
    }
    changePasswordHandler(ev:any){
      this.setState({password:ev.target.value});
    }
    signInHandler(){
      this.props.signIn(this.state.username,this.state.password)
   }
  render(){
    return(
      <>
        {this.state.isUserLoggedIn&&<Redirect to="/clients"/>}
        <LoginWrapper>
          <label htmlFor="username">Username</label>
          <LoginInput loginError={this.state.loginError} type="text" name="username" onChange={this.changeUsernameHandler.bind(this)}/>
          <label htmlFor="password">Password</label>
          <LoginInput loginError={this.state.loginError} type="password" name="password" onChange={this.changePasswordHandler.bind(this)}/>
          <LoginButton onClick={this.signInHandler.bind(this)}>Sign in</LoginButton>
        </LoginWrapper>
      </>
    )
  }
}
const LoginInput= styled('input')<{loginError:boolean}>`
  border: none;
  border-bottom: 1px solid;
  border-bottom-color:${(props)=>props.loginError?'red':"gray"};
  margin-bottom:15px;
  margin-top:6px;
  text-align:center
`
LoginInput.displayName = "LoginInput";

export const LoginButton = styled.button`
  background: none;
  border: 1px solid black;
  border-radius: 5px;
  padding: 5px 0px;
  margin-top: 25px;
`
LoginButton.displayName = "LoginButton"
export const LoginWrapper = styled.div`
  position:absolute;
  left:50%;
  top: 50%;
  transform:translate(-50%,-50%);
  display: flex;
flex-direction: column;
  text-align:center;
  &>label{
    color:gray
  }
`

interface IProps{
  signIn:any,
  user: IUser|null
}
interface IState{
  username:String,
  password:String,
  loginError:boolean,
  isUserLoggedIn:boolean
}
function mapDispatchToProps(dispatch:any){
  return {
    signIn:(username:String,password:String)=>{dispatch(userActions.signIn(username,password))}
  }
}
function mapStateToProps(state:any){
  return {
    user: state.user
  }
}
export default connect(mapStateToProps,mapDispatchToProps)(LoginBox)
