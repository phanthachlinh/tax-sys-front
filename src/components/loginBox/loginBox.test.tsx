import React from 'react';
import Enzyme,{shallow} from 'enzyme';
import {shallowToJson} from 'enzyme-to-json';
import Adapter from 'enzyme-adapter-react-16'
import {LoginBox} from './loginBox'
import { IUser } from '../../ducks/signIn/signIn';
Enzyme.configure({ adapter: new Adapter() });
describe('Check Loginbox content',()=>{
  it('should contain User name label',()=>{
    let newUser
    const component = shallow(<LoginBox signIn={()=>{}} user={null}/>)
    expect(component.contains(
      <label htmlFor="username">
        Username
      </label>
    )).toBeTruthy()
  })
  it('should contain Password Label',()=>{
    const component = shallow(<LoginBox signIn={()=>{}} user={null}/>)
    expect(component.contains(
      <label htmlFor="password">Password</label>
    )).toBeTruthy()
  })
  it('should contain 2 inputs',()=>{
    let newUser
    const component = shallow(<LoginBox signIn={()=>{}} user={null}/>)
    expect(component.find(
      'LoginInput'
    ).length).toBe(2)
  })
  it('should contain A login button',()=>{
    let newUser
    const component = shallow(<LoginBox signIn={()=>{}} user={null}/>)
    expect(component.find(
      'LoginButton'
    ).length).toBe(1)
  })
  it('should contain login button "Sign In" text',()=>{
    let newUser
    const component = shallow(<LoginBox signIn={()=>{}} user={null}/>)
    console.log(component.find(
      'LoginButton'
    ).text())
    expect(component.find(
      'LoginButton'
    ).text()).toBe('Sign in')
  })
  it('should match the snapshot',()=>{

    expect(shallowToJson(shallow(<LoginBox signIn={()=>{}} user={null}/>))).toMatchSnapshot()
  })
  it('should contain a navigate component',()=>{
    let user:IUser = {ID:6,isManager:true}
    expect(shallow(<LoginBox signIn={()=>{}} user={user}/>).find('Redirect').length).toBe(1)
  })
})
