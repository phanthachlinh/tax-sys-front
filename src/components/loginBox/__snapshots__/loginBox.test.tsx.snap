// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Check Loginbox content should match the snapshot 1`] = `
<Fragment>
  <styled.div>
    <label
      htmlFor="username"
    >
      Username
    </label>
    <LoginInput
      loginError={false}
      name="username"
      onChange={[Function]}
      type="text"
    />
    <label
      htmlFor="password"
    >
      Password
    </label>
    <LoginInput
      loginError={false}
      name="password"
      onChange={[Function]}
      type="password"
    />
    <LoginButton
      onClick={[Function]}
    >
      Sign in
    </LoginButton>
  </styled.div>
</Fragment>
`;
